Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,230 @@
+/-- The excess flow of a vertex (the incoming flow minus the outgoing flow). --/
+def excess [Sub R] := f.incoming v - f.outgoing v
+
+end PseudoFlow
+
+section PreFlowDefinition
+
+variable [Preorder R] [AddCommMonoid R] (N : Network V R) (s t : V)
+
+/-- A `PreFlow` from `s` to `t` is a `PseudoFlow` such that all vertices except `s` and `t` have
+non-negative excess, that is, the outgoing flow is at most as much as the incoming flow. -/
+@[ext]
+structure PreFlow extends N.PseudoFlow where
+  excess_nonneg : ∀ v, v ≠ s → v ≠ t → toPseudoFlow.outgoing v ≤ toPseudoFlow.incoming v
+
+namespace PreFlow
+
+variable {N s t}
+
+instance : Zero (N.PreFlow s t) where
+  zero := { (0 : N.PseudoFlow) with excess_nonneg := fun _ _ _ ↦ le_rfl }
+
+/-- The value of a flow is the amount of flow units that leave the source vertex and arrive at the
+sink vertex. In particular, this is equal to the excess of `t`. -/
+def value [Sub R] (f : N.PreFlow s t) := f.excess t
+
+end PreFlow
+end PreFlowDefinition
+
+section FlowDefinition
+
+variable [Preorder R] [AddCommMonoid R] (N : Network V R) (s t : V)
+
+/-- A `Flow` from `s` to `t` is a `PreFlow s t`, but instead of requiring non-negative excess, it
+requires that the excess is 0 (also called _flow conservation_). -/
+@[ext]
+structure Flow extends N.PreFlow s t where
+  conservation : ∀ v, v ≠ s → v ≠ t → toPseudoFlow.outgoing v = toPseudoFlow.incoming v
+  excess_nonneg v hs ht := (conservation v hs ht).le
+
+namespace Flow
+
+instance : Zero (N.Flow s t) where
+  zero := { (0 : N.PreFlow s t) with conservation := fun _ _ _ ↦ rfl }
+
+end Flow
+end FlowDefinition
+
+section SummingLemmas
+variable [Preorder R] [AddCommGroup R] {N : Network V R}
+
+namespace PseudoFlow
+
+variable (f : N.PseudoFlow)
+
+lemma sum_outgoing_eq_sum_incoming : ∑ v, f.outgoing v = ∑ v, f.incoming v := by
+  unfold outgoing incoming
+  rw [Finset.sum_comm]
+
+@[simp]
+lemma sum_excess_zero : ∑ v, f.excess v = 0 := by simp [excess, sum_outgoing_eq_sum_incoming]
+
+end PseudoFlow
+
+namespace Flow
+
+variable {s t : V} (f : N.Flow s t)
+
+lemma excess_source_zero_of_eq (hst : s = t) : f.excess s = 0 := by
+  rw [← f.sum_excess_zero]
+  apply Eq.symm
+  apply Finset.sum_eq_single_of_mem s (Finset.mem_univ _)
+  intro v _ hs
+  simp [PseudoFlow.excess, f.conservation v hs (hst ▸ hs)]
+
+lemma value_zero_of_eq (hst : s = t) : f.value = 0 := by
+  subst t
+  rw [PreFlow.value, f.excess_source_zero_of_eq rfl]
+
+lemma excess_source_eq_neg_excess_sink : f.excess s = -f.excess t := by
+  wlog hst : s ≠ t
+  · rw [not_not] at hst
+    subst t
+    simp [f.excess_source_zero_of_eq]
+  apply eq_neg_of_add_eq_zero_left
+  rw [← f.sum_excess_zero]
+  apply Eq.symm
+  apply Finset.sum_eq_add_of_mem s t (Finset.mem_univ _) (Finset.mem_univ _) hst
+  intro v _ hv
+  simp [PseudoFlow.excess, f.conservation v hv.left hv.right]
+
+lemma excess_sink_eq_neg_excess_source : f.excess t = -f.excess s := by
+  rw [excess_source_eq_neg_excess_sink, neg_neg]
```

Comment:
Lemmas should be simpler on their RHS than on their LHS. If you make sure this is so, then those lemmas can be made simp and have shorter names
```suggestion
@[simp] lemma neg_excess_sink : -f.excess t = f.excess s := by
  wlog hst : s ≠ t
  · rw [not_not] at hst
    subst t
    simp [f.excess_source_zero_of_eq]
  apply eq_neg_of_add_eq_zero_left
  rw [← f.sum_excess_zero]
  apply Eq.symm
  apply Finset.sum_eq_add_of_mem s t (Finset.mem_univ _) (Finset.mem_univ _) hst
  intro v _ hv
  simp [PseudoFlow.excess, f.conservation v hv.left hv.right]

@[simp] lemma neg_excess_source : -f.excess s = f.excess t := by
  rw [← neg_excess_sink, neg_neg]
```

Reply:
Fair point! I wonder if "simpler" should mean "there is no unary minus anymore", or whether it should be "all f.excess terms are normalized to the same vertex" for example `f.excess s`. What do you think? I myself am not sure if either version could cause annoying `simp` behavior

Reply:
Removing the unary minus should be pretty consistently sensible. Locally the version which normalises to `s` should also be fine, since the type of `f` determines what `s` is; but my expectation is that is much more likely to cause weird behaviour elsewhere.
All that to say, I agree with Yaël here.